### ISSUE TITLE
Improve ALPN.ServerProvider to support JDK 9's ALPN

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#625]
Related to #486.

With JDK 8 and `alpn-boot`, the TLS protocol and the cipher where chosen first, and then `ALPN.ServerProvider.select()` was called.
Because the TLS handshake is undergoing, the `SSLEngine.getHandshakeSession()` is not `null`, and can be used in an implementation of `ALPN.ServerProvider.select()` to extract the negotiated TLS protocol and cipher.
`ALPN.ServerProvider.select()` is called only once, because the cipher is already negotiated.

With JDK 9's ALPN, `ALPN.ServerProvider.select()` is called _before_ the TLS handshake begins, so `SSLEngine.getHandshakeSession()` is `null`.
Furthermore, `ALPN.ServerProvider.select()` must be called multiple times to make sure that all the ciphers that are valid for a certain application protocol are selected.
This is needed because the first cipher selected may be an ECDSA cipher, but the certificate in the keystore is not signed using ECDSA. So despite the cipher being valid for that application protocol, is not valid for the certificate, but unfortunately we will only know this when we call `SSLEngine.unwrap()`, which happens _after_ the application protocol negotiation.
Additionally, we need to pass to `ALPN.ServerProvider.select()` not only the list of the client's application protocols, but also the TLS protocol version and the cipher(s) that we want to validate for those protocols.

In summary, in JDK 8 and `alpn-boot`'s `select(List<String> protocols)` is enough because the TLS protocol and the cipher can be extracted from the TLS handshake session.
JDK 9's ALPN needs `select(String tlsProtocol, String cipher, List<String> protocols)` and the semantic to be called multiple times (with zero side effects).

**Comment by sbordet** [comment]
Proposal A.

Modify `ALPN.ServerProvider.select()` so it takes a `ALPN.ServerProvider.Info` object as parameter, where the `Info` class is:

```
class Info {
    final String tlsProtocol;
    final String cipherSuite;
    final List<String> protocols;
}
```

In this way it would be trivial to create an `Info` object for JDK 8 (extracting the TLS protocol and cipher information from the TLS handshake session).

The multiple calls semantic is more delicate though.
Currently, `select()` assumed it's invoked only once, and it sets the negotiated protocol to the one that it returns to the caller.
When calling it multiple times, it may happen that the first time `h2` is selected, then the second time `http/1.1` is selected. The caller will discard `http/1.1` however, but the side effect on `select()` has now set the negotiated protocol to `http/1.1`.

This can be solved in the implementation of `select()` (and therefore no API changes) just by only remembering the first protocol that was returned by `select()`.

Alternatively, it needs the introduction of a `selected()` method that the caller calls with the final decision on the protocol.

**Comment by sbordet** [comment]
The JDK mechanism for ALPN has changed, adding APIs that are closer to `ALPN.ServerProvider` and as such bridging the two mechanisms is now trivial. Implemented in 11d669092a8e60a8122f24e761ad480046635ca0.
